### PR TITLE
(internal): Simplify `org-roam--list-files` implementation.

### PR DESCRIPTION
**Pull request by juergenhoetzel**
Use destructive elisp functions to optimize for speed and memory
usage.

Refs #104

###### Motivation for this change

slight speedup:
Original:
```
ELISP> (benchmark 1000 '(org-roam--list-files org-roam-directory))
"Elapsed time: 29.403994s (6.193575s in 122 GCs)"
```
Commit:
```
ELISP> (benchmark 1000 '(org-roam--list-files org-roam-directory))
"Elapsed time: 27.186161s (5.973992s in 117 GCs)"
```

### CODE REVIEW

**Comment by juergenhoetzel** [comment]
BTW removing `file-truename` for each file would result in a huge speedup:

```
ELISP> (benchmark 1000 '(org-roam--list-files org-roam-directory))
"Elapsed time: 3.529275s (0.933762s in 18 GCs)"
```
Only downside: Symbolic linked org files don't get resolved.

File: org-roam.el [hunk 1]
```diff
@@ -174,22 +174,10 @@ If FILE is not specified, use the current buffer's file-path."
...
-          (cond
-           ((file-directory-p file)
-            (unless (string-match dir-ignore-regexp file)
-              (setq result (append (org-roam--list-files file) result))))
-           ((and (file-readable-p file)
-                 (org-roam--org-file-p file))
-            (setq result (cons (file-truename file) result)))))
-        result)))
+  (let (result)
+    (dolist (file (directory-files-recursively dir "\\.\\(org\\|org.gpg\\)\\'" nil) result)
```
Comment (jethrokuan) [inline org-roam.el:178]: The regexp for this needs to be built from `org-roam-file-extensions`.

**Comment by jethrokuan** [comment]
The elisp implementation of finding these files should be a backup, this is tracked in an old issue #104

**Comment by juergenhoetzel** [comment]
> The regexp for this needs to be built from org-roam-file-extensions.

Thanks: I amended the commit.

> I'm overall in favour of this, but the old code had a dir-ignore-regexp that tries to be smart by ignoring certain hidden directories (like those that start with ., e.g. .git), not sure if the performance is slow when there are such directories in org-roam-directory with a lot of objects.

I was able to remove the `file-truename` bottleneck (from `directory-files-recursively` docstring):


> Each file name appears in the returned list in its absolute form.


This results in a huge performance speedup:

Original Code:

```
ELISP> (benchmark 1000 '(org-roam--list-files org-roam-directory))
"Elapsed time: 21.594839s (7.690849s in 156 GCs)"
```
Commit:

```
ELISP> (benchmark 1000 '(org-roam--list-files org-roam-directory))
"Elapsed time: 6.259151s (1.435158s in 29 GCs)"
```
I would guess the ".git" special handling would not improve performance, because git uses pack-files with corresponding indices for large repos and the elisp code would be more complex.

**Comment by timorunge** [comment]
@juergenhoetzel, @jethrokuan: If you're using tools like e.g. [Syncthing](https://syncthing.net) which has local versioning (the data is stored in a hidden folder) you will be able to see all versioned files. 

The correct RegEx to exclude hidden folders would be something like `^[^.](.*).(?:(org))(?:.gpg)?` but due to my limited lisp knowledge I was not able to get it working...

**Pull request merged**